cpu/kinetis_common/periph/gpio.c

```diff
@@ -24,384 +24,257 @@
+/**
+ * @brief   Mask used to extract the PORT base address from the gpio_t value
+ */
+#define PORT_ADDR_MASK      (0x00007000)
 
 /**
- * @brief Linked list entry for interrupt configurations.
+ * @brief   Mask used to extract the GPIO base address from the gpio_t value
  */
-typedef struct gpio_int_config_entry {
-    struct gpio_int_config_entry* next; /**< pointer to next entry */
-    gpio_cb_t cb;    /**< callback called from GPIO interrupt */
-    void *arg;       /**< argument passed to the callback */
-    uint32_t irqc;   /**< remember interrupt configuration between disable/enable */
-    uint8_t pin;     /**< pin number within the port */
-} gpio_int_config_entry_t;
-
-/* Linked list of interrupt handlers for each port.
- * Using a linked list saves memory when less than 80% of all GPIO pins on the
- * CPU are configured for interrupts, which is true for (almost) all real world
- * applications. */
-static gpio_int_config_entry_t* gpio_interrupts[PORT_NUMOF];
-
-static mutex_t int_config_lock = MUTEX_INIT;
-
-/* Maximum number of simultaneously enabled GPIO interrupts. Each pool entry
- * uses 20 bytes of RAM.
+#define GPIO_ADDR_MASK      (0x000001c0)
+
+/**
+ * @brief   Cleaned up PORT base address
  */
-#ifndef GPIO_INT_POOL_SIZE
-#define GPIO_INT_POOL_SIZE 16
-#endif
+#define PORT_ADDR_BASE      (PORTA_BASE & ~(PORT_ADDR_MASK))
 
-/* Pool of linked list entries which can be used by any port configuration.
- * Rationale: Avoid dynamic memory inside low level periph drivers. */
-gpio_int_config_entry_t config_pool[GPIO_INT_POOL_SIZE];
+/**
+ * @brief   Cleaned up GPIO base address
+ */
+#define GPIO_ADDR_BASE      (GPIOA_BASE & ~(GPIO_ADDR_MASK))
 
-static PORT_Type * const _port_ptrs[] = PORT_BASE_PTRS;
-static GPIO_Type * const _gpio_ptrs[] = GPIO_BASE_PTRS;
+/**
+ * @brief   Kinetis CPUs have 32 pins per port
+ */
+#define PINS_PER_PORT       (32)
 
-static inline uint32_t _port_num(gpio_t dev) {
-    return (uint32_t)((dev & GPIO_PORT_MASK) >> GPIO_PORT_SHIFT);
-}
+/**
+ * @brief   Calculate the needed memory (in byte) needed to save 4 bits per MCU
+ *          pin
+ */
+#define ISR_MAP_SIZE        (GPIO_PORTS_NUMOF * PINS_PER_PORT * 4 / 8)
 
-static inline uint8_t _pin_num(gpio_t dev) {
-    return (uint8_t)((dev & GPIO_PIN_MASK) >> GPIO_PIN_SHIFT);
-}
+/**
+ * @brief   Define the number of simultaneously configurable interrupt channels
+ *
+ * We have configured 4-bits per pin, so we can go up to 16 simultaneous active
+ * extern interrupt sources.
+ */
+#define CTX_NUMOF           (8U)
 
-static inline PORT_Type *_port(gpio_t dev) {
-    return _port_ptrs[_port_num(dev)];
-}
+/**
+ * @brief   Interrupt context data
+ */
+typedef struct {
+    gpio_cb_t cb;
+    void *arg;
+    uint32_t state;
+} isr_ctx_t;
 
-static inline GPIO_Type *_gpio(gpio_t dev) {
-    return _gpio_ptrs[_port_num(dev)];
-}
+/**
+ * @brief   Allocation of memory for each independent interrupt slot
+ */
+static isr_ctx_t isr_ctx[CTX_NUMOF] = {{0}};
```

Comment:
misleading initializer syntax, AFAIK this only specifies the value of the first element of the array, all other elements will be zero-filled, even if the `{0}` is changed to something else...